src/google/protobuf/compiler/csharp/csharp_helpers.h

```diff
@@ -158,6 +158,34 @@ inline bool IsProto2(const FileDescriptor* descriptor) {
   return descriptor->syntax() == FileDescriptor::SYNTAX_PROTO2;
 }
 
+inline bool SupportsPresenceApi(const FieldDescriptor* descriptor) {
+  // Not quite the same as descriptor->is_singular_with_presence() in two ways:
+  
+  // In proto3 we add field presence only for explicitly-optional fields. Message fields
+  // don't get field presence by default, because the user can just compare with a null reference.
+  
+  // We generate Has/Clear for oneof fields in C#.
+  // It's possible that we shouldn't, but stopping doing so would be a breaking change for proto2.
+  // Fortunately the decision is moot for proto3: you can't use "optional" inside a oneof anyway.
+
+  if (IsProto2(descriptor->file())) {
+    // Proto2: every singular field has presence. (Even fields in oneofs.)
+    return !descriptor->is_repeated();
+  } else {
+    // Proto3: only for explictly-optional fields. (Repeated fields can never be explicitly optional,
+    // so we don't need to check for that.)
+    // Currently we can't get at proto3_optional directly, but we can use has_optional_keyword as a surrogate check.    
+    return descriptor->has_optional_keyword();
```

Comment:
Ideally I think we should avoid changing the API based on whether the `optional` keyword is present. Is there a reason that message fields without `optional` shouldn't have `has` and `clear` methods?

Reply:
Yes - it's introducing extra members for no reason. In C# clearing a message field is as simple as saying `foo.SomeField = null;` and likewise checking for presence is either `foo.SomeField != null` or `foo.SomeField is object`. Those are the idiomatic ways of checking for presence - why make the API surface significantly bigger for no benefit?

As above, I'm happy to discuss this more, but would like to make any change in a future PR.

Reply:
How is this different from proto3 optional? Is this CL returning a default instance instead of `null` for unset proto3 optional fields?

Most languages will not have any API change when you add/remove the `optional` keyword for a message field. So I am worried that people will make this change in `.proto` files assuming that it is both API and wire compatible. It would be unfortunate if this could inadvertently break C# users by changing the API.

Reply:
I haven't changed anything about what the property returns. I was just trying to stick to this guideline in the implementation doc:

> For consistency with proto2, proto3 optional fields should use exactly the same API as proto2 optional.

Proto2 message fields have `HasFoo`/`Clear()` members. Personally I don't like that (for the reasons stated earlier) but I was trying to follow the guidelines. Definitely happy to make an exception in this case though.

Reply:
Done (so a proto3 optional message field does *not* have Has/Clear)